File: leetcode/challenge-2021-03/01/01.go
package main

import (
	"fmt"
)

func main() {
	// candyType := []int{1, 1, 2, 2, 3, 3} // A: 3
	// candyType := []int{1, 1, 2, 3} // A: 2
	// candyType := []int{6, 6, 6, 6} // A: 1
	candyType := []int{0, 0, 0, 4} // A: 2

	fmt.Println(distributeCandies(candyType))
}

type void struct{}

func distributeCandies(candyType []int) int {
	maxAllowed := len(candyType) / 2
	c := make(map[int]void)

	for _, t := range candyType {
		c[t] = void{}
	}

	if len(c) < maxAllowed {
		return len(c)
	}

	return maxAllowed
}

// func distributeCandies(candyType []int) int {
// 	sort.Ints(candyType)
// 	maxAllowed := len(candyType) / 2
// 	currType := -1
// 	numTypes := 0

// 	for _, t := range candyType {
// 		if t != currType {
// 			currType = t
// 			numTypes++
// 		}
// 	}

// 	if numTypes < maxAllowed {
// 		return numTypes
// 	}

// 	return maxAllowed
// }

File: leetcode/challenge-2021-03/01/01.rs
use std::collections::HashSet;

impl Solution {

    pub fn distribute_candies(candy_type: Vec<i32>) -> i32 {
        let max_allowed = candy_type.len() / 2;
        let mut c_types = HashSet::new();

        for t in candy_type.iter() {
            c_types.insert(*t);
        }
        
        if c_types.len() < max_allowed {
            return c_types.len() as i32;
        }

        return max_allowed as i32;
    }
}

struct Solution {}

fn main() {
    let candy_type = vec![0, 0, 0, 4];
    println!("{}", Solution::distribute_candies(candy_type));
}